File: collect_all_genotype.py
#!/usr/bin/env python
# coding=utf-8
import re,  sys
import getopt
from collections import defaultdict
from itertools import *
from Bio import SeqIO

def help():
	print '''
	python genotype.py <genotype.list> <output>
	'''
	exit()

if len(sys.argv) < 3:
	help()

dict=['total', 'normal', '12S', '12R', '12C', '12D', '12A', '12V', '13R', '13C', '13D', '13V', '14I', 'nocountable']

all=defaultdict(defaultdict(dict))
all={}
info={}
samples=[]

with open(sys.argv[1], 'r') as LL:
	for line in LL:
		v = line.split('\t')
		sampleID=v[0]
		file = v[1].strip()
		samples.append(sampleID)

		with open(file, 'r') as IN:
			for data in IN.readlines():
				data.strip()
				arr = data.split()
				all.update({arr[0]:{sampleID:{'COUNT':arr[1], 'RATIO':arr[2]}}})

				info[arr[0]]=arr[3]
				print len(all)
		print "------------------", sampleID, "----------------"

with open(sys.argv[2], 'w') as OO:
	#title for output
	OO.write("GT\t")
	for s in samples:
		OO.write(s+"_freq\t"+s+"_ratio\t")
	OO.write("info\n")
	#data 
	for k in dict:
		OO.write(k+"\t")
		for sid in samples:
			if all.get(k):
				if all[k].get(sid):
					count = all[k][sid]['COUNT']
					ratio = all[k][sid]['RATIO']
					OO.write(count+"\t"+ratio+"\t")
				else:
					OO.write("0\t0\t")
			else:
				OO.write("0\t0\t")
		if info.get(k):
			OO.write(info[k]+"\n")
		else:
			OO.write("-\n")

File: collect_all_genotype.pl
use strict;
use warnings;

if(@ARGV<2)
{
	print "perl $0 <genotype.list> <output>	<sampleinfo> <date>\n";
	exit;
}

open SS,$ARGV[2] or die $!;
my %sinfo=();
while(<SS>)
{
	chomp;
	my @a = split /\t/,$_;
	#$sinfo{$a[0]} = $a[1];
	$sinfo{"$a[0]-$a[1]"} = $a[2];
}
close SS;

open OO, ">$ARGV[1]" or die $!;

my @dict=('total', 'normal', 'pathogenic', '12S', '12R', '12C', '12D', '12A', '12V', '13R', '13C', '13D', '13V', '14I', 'uncountable');
my @sample=();
my %h=();
my %info=();

open LL, $ARGV[0] or die $!;
while(<LL>)
{
	chomp;
	my @a = split;
	next unless(exists $sinfo{$a[0]});
	push(@sample, $a[0]);
	open FF, $a[1] or die $!;
	while(<FF>)
	{
		chomp;
		my @b = split /\t/, $_;
		#print STDERR "$b[0]\t$a[0]\t$b[1]\t$b[2]\n";
		$h{$b[0]}{$a[0]}{'C'} = $b[1];
		$h{$b[0]}{$a[0]}{'R'} = $b[2];
		$info{$b[0]} = $b[3];
	}
	close FF;
}
close LL;

for(my $i=0;$i<@dict;$i++)
{
	print OO "#$dict[$i]:$info{$dict[$i]}\n";
}
print OO "\nID\tSample\tDate";
for(my $i=0;$i<@dict;$i++)
{
	print OO "\t$dict[$i]_count";
}
for(my $i=0;$i<@dict;$i++)
{
	print OO "\t$dict[$i]_ratio";
}
print OO "\n";
for(my $x=0;$x<@sample;$x++)
{
	print OO "$sample[$x]\t$sinfo{$sample[$x]}\t$ARGV[3]";;
	for (my $i=0;$i<@dict;$i++ )
	{
		if(exists $h{$dict[$i]}{$sample[$x]}){
			print OO "\t$h{$dict[$i]}{$sample[$x]}{'C'}";
		}else{
			print OO "\t0";
		}
	}
	for(my $j=0;$j<@dict;$j++)
	{
		if(exists $h{$dict[$j]}{$sample[$x]}){
			print OO "\t$h{$dict[$j]}{$sample[$x]}{'R'}%";
		}else{
			print OO "\t0.00%";
		}
	}
	print OO "\n";
}
